src/storm/storage/BitVector.cpp

```diff
@@ -704,90 +753,128 @@ BitVector::const_iterator BitVector::begin() const {
     }
 
-    // For efficiency reasons, we branch on the desired value at this point
-    if (value) {
-        while (startingIndex < endIndex) {
-            // Compute the remaining bucket content.
-            uint64_t remainingInBucket = *bucketIt & mask;
-
-            // Check if there is at least one bit in the remainder of the bucket that is set to true.
-            if (remainingInBucket != 0) {
-                // As long as the current bit is not set, move the current bit.
-                while ((remainingInBucket & (1ull << (63 - currentBitInByte))) == 0) {
-                    ++currentBitInByte;
-                }
+    uint64_t currentBucketIndexOffset = Backward ? endIndex - 1 : startingIndex;
+    uint_fast8_t currentBitInBucket = currentBucketIndexOffset & mod64mask;
+    uint64_t const* bucketIt = dataPtr + (currentBucketIndexOffset >> 6);
+    currentBucketIndexOffset = (currentBucketIndexOffset >> 6 << 6);
 
-                // Only return the index of the set bit if we are still in the valid range.
-                if (startingIndex + currentBitInByte < endIndex) {
-                    return startingIndex + currentBitInByte;
-                } else {
-                    return endIndex;
-                }
-            }
-
-            // Advance to the next bucket.
-            startingIndex += 64;
-            ++bucketIt;
-            mask = -1ull;
-            currentBitInByte = 0;
-        }
+    // Get relevant contents of the first bucket (the one that contains the bit with index currentBucketIndexOffset + currentBitInBucket)
+    uint64_t relevantBitsInBucket;
+    if (Backward) {
+        relevantBitsInBucket = -1ull << (63 - currentBitInBucket);  // 111..111'1'000...000 where the last '1' is at the currentBitInBucket
     } else {
-        while (startingIndex < endIndex) {
-            // Compute the remaining bucket content.
-            uint64_t remainingInBucket = *bucketIt & mask;
-
-            // Check if there is at least one bit in the remainder of the bucket that is set to false.
-            if (remainingInBucket != (-1ull & mask)) {
-                // As long as the current bit is not false, move the current bit.
-                while ((remainingInBucket & (1ull << (63 - currentBitInByte))) != 0) {
-                    ++currentBitInByte;
+        relevantBitsInBucket = -1ull >> currentBitInBucket;  // 000..000'1'111..111 where the first '1' is at the currentBitInBucket
+    }
+    uint64_t currentBucket = Value ? (*bucketIt & relevantBitsInBucket) : (*bucketIt | ~relevantBitsInBucket);
+
+    // Find the right bucket
+    if (currentBucket == (Value ? 0ull : -1ull)) {
+        // The first bucket does not contain a bit with the desired value...
+        do {
+            // Move to next bucket (if there is some)
+            if constexpr (Backward) {
+                if (currentBucketIndexOffset <= startingIndex) {
+                    // No bucket found!
+                    return startingIndex;
```

Comment:
Turns out the doxygen was outdated, good find! 
In backwards mode, `startIndex` serves as the "bit not found" result which is valid because we always return the found index + 1.  I also fixed the `startingIndex >= endIndex` case at the beginning.

Reply:
Great, thanks!